### DISCUSSION TITLE
PolyData changes when you permute the faces circlically

### DISCUSSION THREAD

**Discussion by pkuehne22** [pyvista/pyvista#4348] [Q&A] [answered]
Hi, 

I found something in pyvista which I dont understand. If I define some verts and some faces I expected to get the same polydata result when permuting the faces circlically. But the results changes.  The following shows a minmal working example to reproduce this. 
```py

verts = np.array([[7.476, 0.45, 80.94], [7.146, 0.3, 80.705], [7.286, 0, 80.705], [12.168, 2.312, 84.212], [4.563, 18.608, 84.212], [0, 16.474, 80.94]])

faces1 = np.array([7, 0, 1, 2, 3, 4, 5, 0])
faces2 = np.array([7, 4, 5, 0, 1, 2, 3, 4])
faces3 = np.array([7, 5, 0, 1, 2, 3, 4, 5])
data = pv.PolyData(verts, faces1)
data.plot()
```

I would appreciate every kind of clue. 

Greetings Philip

**Comment by MatthewFlamm** [comment]
It is possible that the PyVista documentation needs to be clearer on this, but you should not be repeating the first point.

https://vtk.org/doc/nightly/html/classvtkPolygon.html#details

> Define the polygon with n-points ordered in the counter- clockwise direction; do not repeat the last point.

```py

verts = np.array([[7.476, 0.45, 80.94], [7.146, 0.3, 80.705], [7.286, 0, 80.705], [12.168, 2.312, 84.212], [4.563, 18.608, 84.212], [0, 16.474, 80.94]])

faces1 = np.array([6, 0, 1, 2, 3, 4, 5])
faces2 = np.array([6, 4, 5, 0, 1, 2, 3])
faces3 = np.array([6, 5, 0, 1, 2, 3, 4])
data = pv.PolyData(verts, faces1)
data.plot()
```

**Reply by dbr-work** [reply]
Actually the above mentioned behaviour persists even when not repeating the last point. I did some further research into the underlying vtk functions and created two vtk files directly using the vtkPackage and the point definitions as given in your answer.
The content of the first file is with the vertices connected starting from the sixth point:

```# vtk DataFile Version 5.1
vtk output
ASCII
DATASET POLYDATA
POINTS 6 double
7.476 0.45 80.94 7.146 0.3 80.705 7.286 0 80.705 
12.168 2.312 84.212 4.563 18.608 84.212 0 16.474 80.94 

POLYGONS 2 6
OFFSETS vtktypeint64
0 6 
CONNECTIVITY vtktypeint64
5 0 1 2 3 4
```
while the content of the second file is with the vertices connected starting from the first point:
```
# vtk DataFile Version 5.1
vtk output
ASCII
DATASET POLYDATA
POINTS 6 double
7.476 0.45 80.94 7.146 0.3 80.705 7.286 0 80.705 
12.168 2.312 84.212 4.563 18.608 84.212 0 16.474 80.94 

POLYGONS 2 6
OFFSETS vtktypeint64
0 6 
CONNECTIVITY vtktypeint64
0 1 2 3 4 5 
```
As far as I know both files should describe the same polygon, however the first file gives an erroneous polygon output as can be seen in the following figure:
![grafik](https://user-images.githubusercontent.com/131985095/234895828-65244659-75dd-4210-af09-26298c3d0bd8.png)
Interestingly, the second file produces the correct polygon:
![grafik](https://user-images.githubusercontent.com/131985095/234895970-ba2d8c09-98bf-4164-9f9b-8bef6e806a21.png)

**Reply by MatthewFlamm** [reply]
You are right.  I think this linked discussion says that while it is possible to store and do some geometric operatins with a non-convex polygon in vtk and thus pyvista, rendering might not work correctly unless it is convex:

https://public.kitware.com/pipermail/vtkusers/2014-February/083146.html

**Comment by darikg** [comment]
I *think* your polygon is 1) not convex and 2) not planar

The following seems to display correctly regardless of cyclic order:

```python
from vtk import vtkContourTriangulator
from pyvista.core.filters import _get_output

normal = data.compute_normals().cell_data['Normals'][0]
data_planar = data.project_points_to_plane(normal=normal)

tfilt = vtkContourTriangulator()
tfilt.SetInputData(data_planar.extract_all_edges())
tfilt.Update()
triangulated = _get_output(tfilt)

triangulated.plot(show_edges=True)
```

**Reply by MatthewFlamm** [reply]
I still get the behavior when the cell is made clearly planar, by setting all z-coordinates to the same value, but still concave.  So I think it is the concavity that is playing the largest role in this behavior.  Triangulating seems to be the answer for rendering purposes of such a polygon.

**Reply by darikg** [reply]
I found that the vtkContourTriangulator didn't work correctly unless the data were replanarized

**Reply by dbr-work** [reply]
I can confirm that the vtkContourTriangulator works like a charm for a planar polygon as suggested in the previous replies. This also applies to the first reply from Matthew pointing towards vtkTriangleFilter, however it eventually also fails if applied to a non-planar polygon. I did some reading about handling non-convex and non-planar polygons in pyVista/vtk as well as in general and found that dealing with this kind of data is not easy with some people recommending using a constrained delaunay filter (didn't work for me yet).
Actually, the points of the polygon in the above example are in one plane (although not axis aligned and only within numerical accuracy). Do you have some ideas how to deal with this kind of polygon, i.e. applying a triangulation? An obvious possibility would be to transform the polygon into a planar one using the normal, however I am afraid that would be computionally intensive for a larger amount of structures.

**Reply by dbr-work** [reply, marked as answer]
Just a short ammendment to my last post. I have just tried using pv.fit_plane_to_points first and project_points_to_plane afterwards to make sure that all points lie as accurately as possible on a single plane. Subsequently, I passed to PolyData to the vtkContourTriangulator. This method results in the correct polygon being rendered. Thank you for the great answers.
A fast method to make sure all points are within the accurary of the given data in a single plane will be the next step to think about.
